src/lsp4clj/json_rpc.clj

```diff
@@ -0,0 +1,115 @@
+(ns lsp4clj.json-rpc
+  "Models LSP JSON-RPC as core.async channels of messages (Clojure hashmaps).
+
+  https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#baseProtocol"
+  (:require
+   [camel-snake-kebab.core :as csk]
+   [camel-snake-kebab.extras :as cske]
+   [cheshire.core :as json]
+   [clojure.core.async :as async]
+   [clojure.java.io :as io]
+   [clojure.string :as string]))
+
+(set! *warn-on-reflection* true)
+
+(defn ^:private read-n-chars [^java.io.InputStream input content-length charset-s]
+  (let [buffer (byte-array content-length)]
+    (loop [total-read 0]
+      (when (< total-read content-length)
+        (let [new-read (.read input buffer total-read (- content-length total-read))]
+          (when (< new-read 0)
+            ;; TODO: return nil instead?
+            (throw (java.io.EOFException.)))
+          (recur (+ total-read new-read)))))
+    (String. ^bytes buffer ^String charset-s)))
+
+(defn ^:private parse-header [line headers]
+  (let [[h v] (string/split line #":\s*" 2)]
+    (when-not (contains? #{"Content-Length" "Content-Type"} h)
+      (throw (ex-info "unexpected header" {:line line})))
```

Comment:
This seems unnecessarily constricting. Should unexpected headers simply be skipped in the spirit of the [robustness principle](https://en.wikipedia.org/wiki/Robustness_principle)?